hummingbot/connector/exchange/bloxroute_openbook/bloxroute_openbook_api_order_book_data_source.py

```diff
@@ -0,0 +1,106 @@
+            """this function is not needed for bloxroute_openbook data source
+                           the request is handled in the _order_book_snapshot func"""
+        )
+
+    async def _order_book_snapshot(self, trading_pair: str) -> OrderBookMessage:
+        """
+        Retrieves a copy of the full order book from the exchange, for a particular trading pair.
+
+        :param trading_pair: the trading pair for which the order book will be retrieved
+
+        :return: the response from the exchange (JSON dictionary)
+        """
+        orderbook: GetOrderbookResponse = await self._provider.get_orderbook(
+            market=trading_pair, limit=1, project=OPENBOOK_PROJECT
+        )
+
+        snapshot_timestamp: float = time.time()
+
+        order_book_message_content = {
+            "trading_pair": trading_pair,
+            "update_id": int(time.time()),
+            "bids": [(bid.price, bid.size) for bid in orderbook.bids],
+            "asks": [(ask.price, ask.size) for ask in orderbook.asks],
+        }
+        return OrderBookMessage(OrderBookMessageType.SNAPSHOT, order_book_message_content, snapshot_timestamp)
+
+    async def listen_for_subscriptions(self):
+        pass
+
+    async def _connected_websocket_assistant(self) -> WSAssistant:
+        raise "connected websocket assistant not yet supported"
+
+    async def _subscribe_channels(self, ws: WSAssistant):
+        """
+        Subscribes to the trade events and diff orders events through the provided websocket connection.
+        :param ws: the websocket assistant used to connect to the exchange
+        """
+        raise "subscribe channels not yet supported"
+
+    def _channel_originating_message(self, event_message: Dict[str, Any]) -> str:
+        raise Exception("Bloxroute Openbook does not use `_channel_originating_message`")
+
+    async def _process_websocket_messages(self, _: WSAssistant):
+        orderbook_queue = self._message_queue[self._snapshot_messages_queue_key]
+        async for orderbook_event in self._orderbook_stream:
+            orderbook_queue.put_nowait(orderbook_event.to_dict(include_default_values=True))
+
+    async def _parse_order_book_snapshot_message(self, raw_message: Dict[str, Any], message_queue: asyncio.Queue):
+        order_book_message: OrderBookMessage = BloxrouteOpenbookOrderBook.snapshot_message_from_exchange(
+            raw_message,
+            time.time(),
+        )
+        message_queue.put_nowait(order_book_message)
+
+    async def _parse_order_book_diff_message(self, raw_message: Dict[str, Any], message_queue: asyncio.Queue):
+        raise Exception("Bloxroute Openbook does not use orderbook diffs")
+
+    async def _parse_trade_message(self, raw_message: Dict[str, Any], message_queue: asyncio.Queue):
+        raise Exception("Bloxroute Openbook does not use trade updates")
+
+    async def listen_for_order_book_diffs(self, ev_loop: asyncio.AbstractEventLoop, output: asyncio.Queue):
```

Comment:
this one makes sense to raise an Exception for; to clearly identify that this function doesn't work for Trader API

Reply:
I believe all the "listen_for_..." functions are called in the `start` function of the base class in`order_book_tracker.py`, so I just pass them